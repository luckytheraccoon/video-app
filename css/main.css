html{overflow-x:hidden}body,html{margin:0;padding:0;font-family:'Varela Round',sans-serif;font-size:14px;width:100%;height:100%}#root{margin:0 auto;max-width:1000px;min-width:260px;height:100%}#root>div{width:100%;height:100%}a{cursor:pointer;color:#317795}a:link,a:visited{text-decoration:none!important;font-weight:400}a:focus,a:hover{color:#23527c}.div-loading{padding-top:10px;text-align:center}.div-main{display:grid;width:100%;height:100%}.div-maincontent-wrapper{align-self:start;margin:10px;grid-area:maincontent}.div-search-form{margin-bottom:10px;flex:1 100%;justify-content:space-between;flex-wrap:wrap;align-self:start}.div-search-form>input.search-box{flex-grow:1;margin-right:2px}.div-search-form>input.submit-button{flex-shrink:1}.div-user-block{text-align:center}.div-user-block>.div-user-photo{font-size:5em;line-height:.1em;padding:10px}.div-main-menu{background:#f1f1f1;z-index:1;flex-direction:column;justify-content:space-between;min-width:150px}.div-main-menu>div{flex-shrink:1}.div-main-menu>.div-main-menu-menu{flex-grow:1;padding:0 5px 5px 5px}.div-main-menu>.div-main-menu-menu>button{width:100%;margin-top:5px}.div-main-menu>.div-main-menu-logout{padding:5px}.div-main-menu>.div-main-menu-logout>button{width:100%}.elementHidden{display:none}.div-content-list,.div-main-menu.elementVisible,.div-search-form.elementVisible,.div-side-content-list{display:flex}.div-content-list{flex-flow:wrap}.div-content-detail,.div-content-list,.div-side-content-list{position:relative}.div-content-detail>div>div,.div-content-list>div>div,.div-side-content-list>div>div{margin:0 0 10px 0}.div-content-detail>div>div.title,.div-content-list>div>div.title,.div-side-content-list>div>div.title{font-size:.9em;margin:0}.div-content-detail>div>div.description,.div-content-list>div>div.description,.div-side-content-list>div>div.description{font-size:.8em}.div-side-content-list{flex-direction:column;padding-left:10px;flex:1 20%}.div-content-detail>div>div>div.title{margin-top:10px;font-size:1.2em;font-weight:700}.div-footer{grid-area:footer;display:flex;justify-content:space-between;flex-wrap:wrap;background-color:#f1f1f1;padding:25px;color:#888}.div-footer>div{margin:5px;flex-shrink:1}.div-footer>div:last-child{text-align:right;flex-grow:1}.video-iframe{position:absolute}.thumbnail-large>a>img{width:100%;max-width:275px;cursor:pointer}.thumbnail-medium>a>img{width:100%;cursor:pointer}.glyphicon{cursor:pointer}a>.glyphicon{color:#fff}.go-back-button,.more-results-button,.more-results-button-no-click{width:100%!important;padding:10px!important;background:#f1f1f1;border:1px solid #ccc;margin-bottom:10px;text-align:center;cursor:pointer;color:#000}.go-back-button>.glyphicon,.more-results-button-no-click>.glyphicon,.more-results-button>.glyphicon{color:#000}.go-back-button:hover,.more-results-button-no-click:hover,.more-results-button:hover{border:1px solid #999;background:#f5f5f5}.more-results-button-no-click:hover{background:#f1f1f1;border:1px solid #f1f1f1}@media (min-width:768px) and (max-width:991px){.div-main{grid-template-rows:[maincontent] 9fr [footer] 1fr;grid-template-areas:"maincontent-left maincontent" "footer footer";grid-template-columns:[left] 1fr [right] 9fr}.div-maincontent-wrapper{display:flex;flex-flow:row wrap}.div-side-content-list{padding:10px 0}.div-side-content-list .thumbnail-medium{float:left;margin-right:10px}.div-content-detail,.div-content-list{flex:1 100%}.div-content-detail.elementHidden,.div-content-list.elementHidden{display:flex}.div-content-detail>div>.thumbnail-large,.div-content-list>div>.thumbnail-large{float:left;margin-right:10px}.div-content-detail>div>div{flex:1 100%}.div-content-list>div{flex:1 100%}.div-search-form{grid-area:searchbar}.div-search-form.elementHidden{display:flex}.div-main-menu{grid-area:maincontent-left}.div-main-menu.elementHidden{display:flex}.div-mobile-top-menu-center,.div-mobile-top-menu-left,.div-mobile-top-menu-right{display:none}}@media (min-width:992px){.div-main{grid-template-rows:[maincontent] 9fr [footer] 1fr;grid-template-areas:"maincontent-left maincontent" "footer footer";grid-template-columns:[left] 1fr [right] 9fr}.div-maincontent-wrapper{display:flex;flex-flow:row wrap}.div-content-detail,.div-content-list{flex:1 80%}.div-content-detail.elementHidden,.div-content-list.elementHidden{display:flex}.div-content-detail>div>div{flex:1 100%}.div-content-list>div{padding:0 5px;width:33.3%}.div-search-form{grid-area:searchbar}.div-search-form.elementHidden{display:flex}.div-main-menu{grid-area:maincontent-left}.div-main-menu.elementHidden{display:flex}.div-mobile-top-menu-center,.div-mobile-top-menu-left,.div-mobile-top-menu-right{display:none}}@media (max-width:767px){.div-main{grid-template-rows:[header] 60px [maincontent] 9fr [footer] 1fr;grid-template-areas:"mobile-top-menu-left mobile-top-menu-center mobile-top-menu-right" "maincontent maincontent maincontent" "footer footer footer";grid-template-columns:[left] 2fr [center] 6fr [right] 2fr}.go-back-button{display:none}.div-side-content-list{padding:10px 0;flex-direction:column;flex:1 100%}.div-side-content-list .thumbnail-medium{width:40%;float:left;margin-right:10px}.div-content-detail,.div-content-list{flex:1 100%}.div-content-detail>div,.div-content-list>div{flex:1 100%}.div-content-detail>div>.thumbnail-large,.div-content-list>div>.thumbnail-large{width:40%;float:left;margin-right:10px}.div-content-detail>div>.description,.div-content-detail>div>.title,.div-content-list>div>.description,.div-content-list>div>.title,.div-side-content-list>div>.description,.div-side-content-list>div>.title{width:55%;float:left}.div-main-menu{grid-area:maincontent;padding:10px}.div-mobile-top-menu-left{grid-area:mobile-top-menu-left;text-align:left;padding:10px 0 4px 14px;font-size:2em;background-color:#b2b200;position:relative;color:#fff}.div-mobile-top-menu-center{grid-area:mobile-top-menu-center;text-align:center;padding:10px 0 4px 0;font-size:2em;background-color:#b2b200;position:relative;color:#fff}.div-mobile-top-menu-right{grid-area:mobile-top-menu-right;text-align:right;padding:10px 14px 4px 0;font-size:2em;background-color:#b2b200;position:relative;color:#fff}}@media (max-width:390px){.div-main{grid-template-rows:[header] 40px [maincontent] 9fr [footer] 1fr}.div-mobile-top-menu-center,.div-mobile-top-menu-left,.div-mobile-top-menu-right{font-size:1.2em}}