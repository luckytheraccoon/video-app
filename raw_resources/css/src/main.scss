@mixin full-span() {
	width:100%;
	height:100%;
}

@function str-replace($string, $search, $replace: '') {
  $index: str-index($string, $search);

  @if $index {
    @return str-slice($string, 1, $index - 1) + $replace + str-replace(str-slice($string, $index + str-length($search)), $search, $replace);
  }

  @return $string;
}

@mixin prefix($declarations, $prefixes: ()) {
  @each $property, $value in $declarations {
    @each $prefix in $prefixes {
      @if type_of($value)==string  {
        $value: str-replace($value, '-', ',');
        $value: str-replace($value, '_', ' ');
      } 
      #{'-' + $prefix + '-' + $property}: $value;
    }

    // Output standard non-prefixed declaration
    @if type_of($value)==string  {
      $value: str-replace($value, '-', ',');
      $value: str-replace($value, '_', ' ');
    }
    #{$property}: $value;
  }
}
html {
	overflow-x: hidden;
}
html, body {
	margin:0;
	padding:0;
	font: {
		family: 'Varela Round', sans-serif;
		size:14px;
	}
	@include full-span();
}
#root {
	margin:0 auto;
	max-width:1000px;
	min-width:260px;
	height:100%;
	&>div {
		@include full-span();
	}
}
a {
	cursor:pointer;
	color: #317795;
	&:link, &:visited {
	    text-decoration: none !important;
	    font-weight: normal;
	}

	&:hover, &:focus {
	    color: #23527c;
	}

}

@mixin div-main-grid() {
	display: grid;
	@include full-span();
}

@mixin div-header() {
	background-color:#B2B200;
	position: relative;
	color: white;
}

.div-loading {
	padding-top:10px;
	text-align: center;
}

.div-main {
	@include div-main-grid();
}

.div-maincontent-wrapper {
	align-self: start;
	margin: 10px;
	grid-area: maincontent;
}

.div-search-form {
	margin-bottom: 10px;
    flex: 1 100%;
	justify-content: space-between;
	flex-wrap: wrap;
	align-self: start;
	&>input.search-box {
		flex-grow: 1;
		margin-right:2px;
	}
	&>input.submit-button {
		flex-shrink: 1;
	}
}

.div-user-block {
	text-align:center;
	&>.div-user-photo {
		font-size: 5em;
		line-height: 0.1em;
		padding: 10px;
	}
}

.div-main-menu {
	background: #f1f1f1;
    z-index: 1;
    flex-direction: column;
    justify-content: space-between;
	min-width: 150px;
	&>div {
		flex-shrink: 1;
	}

	&>.div-main-menu-menu {
		flex-grow: 1;
		padding:0 5px 5px 5px;
		&>button {
			width: 100%;
			margin-top:5px;
		}
	}

	&>.div-main-menu-logout {
		padding:5px;
		&>button {
			width: 100%;
		}
	}
}

.elementHidden {
	display:none; 
}
.div-side-content-list,
.div-content-list,
.div-search-form.elementVisible,
.div-main-menu.elementVisible {
	display: flex;
}

.div-content-list {
	flex-flow:wrap;
}

.div-side-content-list,
.div-content-detail,
.div-content-list {
	position:relative;
	&>div>div {
		margin:0 0 10px 0;
	}

	&>div>div.title {
		font-size:0.9em;
		margin:0;
	}

	&>div>div.description {
		font-size:0.8em;
	}
}

.div-side-content-list {
	flex-direction:column;
	padding-left: 10px;
	flex: 1 20%;
}

.div-content-detail {
	&>div>div>div.title {
		margin-top:10px;
		font-size:1.2em;
		font-weight: bold;
	}
}
.div-footer {
	grid-area: footer;
	display: flex;
	justify-content: space-between;
	flex-wrap: wrap;
	background-color:#f1f1f1;
	padding:25px;
	color:#888;
	&>div {
		margin:5px;
		flex-shrink:1;
	}
	&>div:last-child {
		text-align: right;
		flex-grow:1;
	}
}
.video-iframe {
	position:absolute;
}

.thumbnail-large>a>img {
	width:100%;
	max-width: 275px;
	cursor: pointer;
}

.thumbnail-medium>a>img {
	width:100%;
	cursor: pointer;
}
.glyphicon {
	cursor: pointer;
}
a>.glyphicon {
	color:white;
}


.more-results-button,
.go-back-button,
.more-results-button-no-click {
	width: 100%!important;
	padding: 10px!important;
	background: #f1f1f1;
	border:1px solid #CCC;
	margin-bottom: 10px;
	text-align: center;
	cursor: pointer;
	color:black;

	&>.glyphicon {
		color:black;
	} 

	&:hover {
		border:1px solid #999;
		background: #f5f5f5;
	}

}

.more-results-button-no-click {
	&:hover {
		background: #f1f1f1;
		border:1px solid #f1f1f1;
	}
}

/*
	Large Tablets / Laptops
*/
@media (min-width:768px) and (max-width:991px) {
    .div-main {
		grid-template: {
			rows: [maincontent] 9fr [footer] 1fr;
			areas: 
			"maincontent-left maincontent"
			"footer footer";
			columns: [left] 1fr [right] 9fr;
		}
	}
	.div-maincontent-wrapper {
		display:flex;
		flex-flow: row wrap;
	}
	.div-side-content-list {
		padding:10px 0;
		& .thumbnail-medium {
			float:left;
			margin-right:10px;
		}
	}
	.div-content-list, 
	.div-content-detail {
		flex: 1 100%;

		&.elementHidden {
			display: flex;
		}
		
		&>div>.thumbnail-large {
			float:left;
			margin-right:10px;
		}
	}
	.div-content-detail {
		&>div>div {
			flex: 1 100%;
		}
	}
	.div-content-list {
		&>div {
			flex: 1 100%;
		}
	}
	.div-search-form {
		grid-area: searchbar;
		&.elementHidden {
			display:flex;
		}
	}

	.div-main-menu {
		grid-area: maincontent-left;  
		&.elementHidden {
			display:flex;
		}
	}

	.div-mobile-top-menu-left,
	.div-mobile-top-menu-center,
	.div-mobile-top-menu-right {
		display:none
	}
}
/*
	Laptops / Desktops
*/
@media (min-width:992px)  {
    .div-main {
		grid-template: {
			rows: [maincontent] 9fr [footer] 1fr;
			areas: 
			"maincontent-left maincontent"
			"footer footer";
			columns: [left] 1fr [right] 9fr;
		}
	}
	.div-maincontent-wrapper {
		display:flex;
		flex-flow: row wrap;
	}
	.div-content-detail,
	.div-content-list {
    	flex: 1 80%;
		&.elementHidden {
			display: flex;
		}
	}
	.div-content-detail {
		&>div>div {
			flex: 1 100%;
		}
	}
	.div-content-list {
		&>div {
			padding: 0 5px;
			width: 33.3%;
		}
	}

	.div-search-form {
		grid-area: searchbar;
		&.elementHidden {
			display:flex;
		}
	}

	.div-main-menu {
		grid-area: maincontent-left;  
		&.elementHidden {
			display:flex;
		}
	}

	.div-mobile-top-menu-left,
	.div-mobile-top-menu-center,
	.div-mobile-top-menu-right {
		display:none
	}
}

/*
	Small Tablets
*/
@media (max-width:767px) {
    .div-main {
		grid-template: {
			rows: [header] 60px [maincontent] 9fr [footer] 1fr;
			areas: 
			"mobile-top-menu-left mobile-top-menu-center mobile-top-menu-right"
			"maincontent maincontent maincontent"
			"footer footer footer";
			columns: [left] 2fr [center] 6fr [right] 2fr;
		}
	}

	.go-back-button {
		display:none;
	}
	
	.div-side-content-list {
		padding:10px 0;
		flex-direction:column;
    	flex: 1 100%;
		& .thumbnail-medium {
			width: 40%;
			float:left;
			margin-right:10px;
		}
	}
	
	.div-content-detail,
	.div-content-list {
    	flex: 1 100%;
		
		&>div {
			flex: 1 100%;
		}
		&>div>.thumbnail-large {
			width: 40%;
			float:left;
			margin-right:10px;
		}
	}

	.div-side-content-list,
	.div-content-detail,
	.div-content-list {
		&>div>.title,
		&>div>.description {
			width: 55%;
			float:left;
		}
	}

	.div-main-menu {
		grid-area: maincontent;    
		padding: 10px;
	}

	@mixin div-mobile-top-menu() {
		font-size:2em;
		@include div-header();
	}
	.div-mobile-top-menu-left {
		grid-area: mobile-top-menu-left;
		text-align:left;
		padding: 10px 0 4px 14px;
		@include div-mobile-top-menu();
	}
	.div-mobile-top-menu-center {
		grid-area: mobile-top-menu-center;
		text-align:center;
		padding: 10px 0 4px 0;
		@include div-mobile-top-menu();
	}
	.div-mobile-top-menu-right {
		grid-area: mobile-top-menu-right;
		text-align:right;
		padding: 10px 14px 4px 0;
		@include div-mobile-top-menu();
	}
}

/*
	Smartphones
*/
@media (max-width:390px) {
	
    .div-main {
		grid-template-rows: [header] 40px [maincontent] 9fr [footer] 1fr;
	}
	.div-mobile-top-menu-left,
	.div-mobile-top-menu-center,
	.div-mobile-top-menu-right {
		font-size:1.2em;
	}
}