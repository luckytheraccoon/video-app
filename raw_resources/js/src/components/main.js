import React from "react";
//custom modules:
//this is where we keep some commonly used components, i like to keep them separate from the larger components for easier access and readability
import { Glyphicon } from "../components/common"; 

//this is where we keep the mobile controls, each page can have different controls
import mobileMenuItems from "../data/mobileTopMenu"; 
 
//this is where we keep the main menu controls, each page can have different controls
import mainMenuItems from "../data/mainMenu"; 

//this is where we keep the pages, each page can have its own output or it can depend on a content object
//the content objects can be obtained from within the mainPages module which imports the mainContent module
import fetchPage from "../data/mainPages"; 

//helper methods that execute general tasks
import { getUrlParameter, buildApiUrl } from "../helpers/common"; 

/**
 * The main container is the main wrapper component that's responsible for 
 * controlling which page (and inner content) should be displayed
 * 
 * the changeContent method is solely responsible for the actual changing of pages and their content
 * a page is a "section" with a title and potential specific controls
 * a content is whatever is to be displayed within each page, it is possible to list content items or visualize a single item
 */  
class MainContainer extends React.PureComponent {

    constructor(props) {
        super(props);
        //binding events, more info on each at their declaration
        this.toggleElement = this.toggleElement.bind(this);
        this.requestList = this.requestList.bind(this);
        this.requestItem = this.requestItem.bind(this);
        this.loadMoreResults = this.loadMoreResults.bind(this);
        this.addToDocumentTitle = this.addToDocumentTitle.bind(this);
        this.moreResultsAction = this.moreResultsAction.bind(this);
        this.pageRequestBuilder = this.pageRequestBuilder.bind(this);

        //if a get request exists with the variable "video" in it, capture the value
        this.getRequestVideo = getUrlParameter('video');

        //the content array is where our current content is sitting, waiting to get rendered somehow
        this.contentArray = [];
        //how many items should we load per request?
        this.maxItemsPerPage = 9;
        //the main content page object that's currently going to be rendered as the central scene
        this.centerPage = {};
        //overlay related css classes for when an element should be visible or invisible due to an overlay being present or not
        this.visibleClass = "elementVisible";
        this.hiddenClass = "elementHidden";
        //the title of the window
        this.DocumentTitle = "OLX Challenge"; 

        this.state = {
            //this represents a content item
            contentId:null, 
            //the default page when the user opens the app
            pageId:"video-list", 
            //this is the default page title
            pageTitle:null, 
            //the page title before it change since the last time, this is useful for when the user navigates from A to B then goes back to A
            previousPageTitle:null,
            //is the mobile menu visible or not?
            mobileMenu:false, 
            //is the search box visible in the mobile menu or not?
            mobileSearch:false, 
            //when true, the content has done fetching from server
            contentLoaded:false,
            //what's the currently searched term?
            currSearchTerm:null, 
            //the current page that's been requested, starting at one
            currRequestPage:1, 
            //how many items have been loaded so far?
            currLoadedItems:null, 
            //what's the max items the server found for the current request?
            maxContentItems:null,
            //how many items did the last request return?
            lastRequestItemsCount:null
        };
    }

    /**
     * Method to set the window title before anything else.
     */
    componentWillMount() {
        this.setDocumentTitle(this.DocumentTitle);
    }

    /**
     * Initial request once this main component has done mounting for the first time.
     */
    componentDidMount() {
        this.pageRequestBuilder(false);
    }

    /**
     * Creates the request system for when the user either enters the app or requests a new page from the current results.
     * 
     * @param {bool} newPageRequest - If this is true, user is requesting a new page for the current results. 
     *                                If false, he's requesting a whole new page, for example the first time the user comes in.
     */
    pageRequestBuilder(newPageRequest) {
        if(this.getRequestVideo) {
            if(newPageRequest === true) {
                this.requestList(this.state.currSearchTerm, true, false);
            } else {
                this.requestList(null, false, false);
            }

            this.requestItem(this.getRequestVideo);

        } else {
            if(newPageRequest) {
                this.requestList(this.state.currSearchTerm, true, true);
            } else {
                this.requestList(null, false, true);
            }
        }
    }

    /**
     * Adds to the current window title
     * 
     * @param {string} newTitle - What to concatenate to the title
     */
    addToDocumentTitle(newTitle) {
        this.setDocumentTitle(this.DocumentTitle + " - " + newTitle);
    }

    /**
     * Sets the window title
     * 
     * @param {string} title - What to set the window title to
     */
    setDocumentTitle(title) {
        document.title = title;
    }
    
    /** 
     * For when a user requests more items by clicking the "load more" button
     * Its just a shortcut for another method, but can be used to change the button's style, for example.
     */
    loadMoreResults() {
        this.pageRequestBuilder(true);
    }

    /**
     * returns the loadMoreResults action only when its possible to load more results, if not, returns null
     */
    moreResultsAction() {
        //can more results be loaded?
        //this controls if the load more button should appear or not
        let moreResultsAction = this.loadMoreResults; //yes

        if(this.state.currLoadedItems >= this.state.maxContentItems) {
            moreResultsAction = null; //no
        }

        return moreResultsAction;
    }

    /**
     * This method is single-handedly responsible for deciding what to render on the main scene.
     * Previously this was in the component's render method but that was not very pretty. 
     */
    getSceneToRender() {

        //a bit of mumbo jumbo going on here:
        //toggle visibility of elements based on the state of said elements
        let mobileMenuVisibleClass = this.state.mobileMenu ? this.visibleClass : this.hiddenClass;
        let mobileSearchVisibleClass = this.state.mobileSearch && !this.state.mobileMenu  ? this.visibleClass : this.hiddenClass;
        //the main scene should be hidden when the mobile menu appears
        let pageVisibleClass = this.state.mobileMenu ? this.hiddenClass : this.visibleClass;

        let mobileIcons, pageTitle, centerPageComponent, sideList, sideListComponent;

        pageTitle = this.centerPage.title;
        centerPageComponent = this.centerPage.component;
        
        //only fetch the side video list when user is in the video-detail page
        if(this.state.pageId === "video-detail") {
            sideList = fetchPage(this.contentArray, "side-video-list", this.state.contentId, this.moreResultsAction);
            sideListComponent = sideList.component;
        }

        mobileIcons = mobileMenuItems(this.state.pageId, this.toggleElement, this.state.pageTitle, this.state.mobileMenu);
        
        return (
            <div className="div-main">
                {mobileIcons}
                <MainMenu className={mobileMenuVisibleClass}/>
                <div className={"div-maincontent-wrapper " + pageVisibleClass }>
                    <Searchform currRequestPage={this.state.currRequestPage} onSubmit={this.requestList} className={mobileSearchVisibleClass} />
                    {centerPageComponent}
                    {sideListComponent}
                </div>
                <Footer />
            </div>
        );
    }

    /**
     * Whenever a request is done to get more content.
     * Takes two arguments, one asks for a new search, the other simply requests a new page for the current search term.
     * 
     * @param {string} requestSearchTerm - A string to send to the server as a search parameter.
     * @param {bool} requestNextReqPage - If true, does not reset counters and does not redirect the user, simple gets next page. If false, always goes back to first page.
     * @param {bool} setPageWhenDone - If true, when this list is done fetching from the server, react can render the page. If false, don't render a page.
     */
    requestList(requestSearchTerm, requestNextReqPage, setPageWhenDone) {

        //if the search term changed, then we should start at page one
        //otherwise, just increment the page number and stack more items onto the contentArray
        //this also decides what state to change, so we dont change everything for no reason

        //the current requested page, we need to define this so we can increment it or reset it
        let currReqPage = this.state.currRequestPage;
        //itemcounter is used to compare with max results, if item counter is higher or equal to max results, then should not be possible to load more
        let itemIndexCounter = this.state.currLoadedItems;
        //stateToChange is used to i dont have to setState multiple times, which is saving some code lines
        let stateToChange = {};

        //i tend to stay away from if-else, but this is for readability:
        //if the search term is different from the previous one, then the user needs new data
        //if requestNextReqPage is false, then the user didnt click the "load more" button, he did a search 
        //otherwise he clicked the "load more" button and that means we only have to get a new page instead of resetting stuff
        if(this.state.currSearchTerm != requestSearchTerm || requestNextReqPage === false) {
            //set the new search term
            stateToChange.currSearchTerm = requestSearchTerm;
            //empty the contentArray const (remember, we can't redefine a const)
            for (var i = this.contentArray.length; i > 0; i--) {
                this.contentArray.pop();
            }
            //these are some "reset" instructions:
            //reset the counter, pageId, page title and set the new search term, reset the current request page
            itemIndexCounter=0;
            stateToChange.currLoadedItems = itemIndexCounter;
            currReqPage=1;
        } else {
            //increment the request page so the user gets a new fresh list of items
            currReqPage++;
        }
        
        //update state with current requested page
        stateToChange.currRequestPage = currReqPage;

        //call helper function to build the URL to fetch data from server
        let apiUrl = buildApiUrl(requestSearchTerm, currReqPage, this.maxItemsPerPage, null);

        //finally execute ajax request, process the json response from API
        fetch(apiUrl).then(function (response) {
            return response.json();
        }).then(function (response) {

            let pageId = "video-list";

            //map each item onto an array of content 
            response.resource.map((item) => {
                item.index = itemIndexCounter;
                this.contentArray[item.id] = item;
                itemIndexCounter++;
            });
            
            //more state updates after processing the list of content
            //notify app that content has loaded and set a page only if we pass true on render page when done
            if(setPageWhenDone === true) {
                this.setPage(pageId);
                stateToChange.contentLoaded = true;
                stateToChange.pageId = pageId;
                stateToChange.pageTitle = this.centerPage.title;
            }
            
            stateToChange.currLoadedItems = itemIndexCounter;
            stateToChange.maxContentItems = response.total;
            
            //update state
            this.setState(stateToChange);
            
        }.bind(this)
        ).catch(function() {
        });
    }

    /**
     * Requests a single item from the server. 
     * Adds to the contentArray so the content get be fetched.
     * 
     * @param {string} contentId - An identifier for a single item.
     */
    requestItem(contentId) {

        //call helper function to build the URL to fetch data from server
        let apiUrl = buildApiUrl(null, null, null, contentId);
        //finally execute ajax request, process the json response from API
        fetch(apiUrl).then(function (response) {
            return response.json();
        }).then(function (response) {

            let pageId = "video-detail";

            //map the obtained item onto an array of content 
            let item = response.resource;
            this.contentArray[item.id] = item;

            this.setPage(pageId, item.id);
            
            this.setState({
                pageId: pageId,
                contentId: item.id,
                contentLoaded: true,
                pageTitle: this.centerPage.title
            });

        }.bind(this)
        ).catch(function() {
        });
    }

    /**
     * Set the page to an object that can be rendered later.
     * 
     * @param {string} pageId - The page Id according to mainPages module.
     * @param {string} contentId - The content to load within the page.
     */
    setPage(pageId, contentId) {
        this.centerPage = fetchPage(this.contentArray, pageId, contentId, this.moreResultsAction, this.addToDocumentTitle);
        this.addToDocumentTitle(this.centerPage.title);
    }

    /**
     * This is used to toggle something into/outof visibility, like an overlay or an element.
     * 
     * When an element is toggled into visibility, it can redefine the page title.
     * 
     * @param {string} stateEntry - The state entry that controls the toggle state.
     * @param {string} newPageTitle - The page title for when toggle is "visible".
     */
    toggleElement(stateEntry, newPageTitle) {
        let stateToChange = {};
        if(this.state[stateEntry]) {
            stateToChange[stateEntry]=false;
            stateToChange.pageTitle=this.state.previousPageTitle;
        } else {
            stateToChange[stateEntry]=true;
            stateToChange.pageTitle=newPageTitle;
            stateToChange.previousPageTitle=this.state.pageTitle;
        }
        this.setState(stateToChange);
    }

    /**
     * Where most of the magic actually happens.
     * Pretty much on rendering the scene, we decide which elements to show or hide based on specific conditions.
     */
    render() {

        //get the page to show, can be a list or a detail view
        //while the content is loading...
        let scene = <div className="div-loading"> Content is loading ... </div>;
        //once the content has finished loading:
        if(this.state.contentLoaded) {
            scene = this.getSceneToRender();
        }

        return scene;
    }
}
/**
 * Component that renders the search form.
 * Invokes ManContainer callback on submit.
 */
class Searchform extends React.PureComponent {

    constructor(props) {
        super(props);
        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.state = {value: ''};
    }

    /**
     * Keep the search box content in state for usage.
     * 
     * @param {object} event - Event info.
     */
    handleChange(event) {
        this.setState({value: event.target.value});
    }

    /**
     * Whenever the form is submitted, don't actually refresh the page and call a MainContainer callback.
     * 
     * @param {object} event - Event info.
     */
    handleSubmit(event) {
        this.props.onSubmit(this.state.value, false, true);
        event.preventDefault();
    }

    render() {
        return(
            <form className={"div-search-form " + this.props.className} onSubmit={this.handleSubmit}>
                <input className="search-box" type="text" value={this.state.value} onChange={this.handleChange} placeholder="What are you looking for?"/>
                <input className="submit-button" type="submit" value="Search" />
            </form>
        );
    }
}

/**
 * This renders the left menu.
 * 
 * @param {object} props - The component properties from within the react context.
 */
function MainMenu(props) {
    //gets the menu buttons from an "external" source
    const menuItems = mainMenuItems().map((item) => {
        return <button key={item.id}>{item.label}</button>;
    });
    return (
        <div className={"div-main-menu " + props.className}>
            <UserBlock />
            <div className="div-main-menu-menu">
                {menuItems}
            </div>
            <div className="div-main-menu-logout">
                <button>Logout</button>
            </div>
        </div>
    );
}

/**
 * Simple component that renders the user block.
 */
function UserBlock() {
    return(
        <div className="div-user-block">
            <div className="div-user-photo"><Glyphicon iconSuffix="user" /></div>
            <div className="div-user-name">User Name</div>
        </div>
    );
}

/**
 * Simple component that renders the footer.
 * Possible todo: implement external source for the content.
 */
function Footer() {
    return (
        <div className="div-footer">
            <div>Footer 1</div>
            <div>Footer 2</div>
            <div>Footer 3</div>
            <div>OLX Copyright 2017</div>
        </div>
    );
}


export { MainContainer, MainMenu, Searchform };