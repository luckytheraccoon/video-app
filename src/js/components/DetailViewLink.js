import React from "react";
/**
 * Just a link that will send the user to the video detail.
 * 
 * @param {object} props - The component properties from within the react context.
 */
export default (props) => {
    return <a href={"?video=" + props.contentId}>{props.children}</a>;
}