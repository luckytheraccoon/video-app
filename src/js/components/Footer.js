import React from "react";
/**
 * Simple component that renders the footer.
 * Possible todo: implement external source for the content.
 */
export default () => {
    return (
        <div className="div-footer">
            <div>Footer 1</div>
            <div>Footer 2</div>
            <div>Footer 331</div>
            <div></div>
        </div>
    );
}