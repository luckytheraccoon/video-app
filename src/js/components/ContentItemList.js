import React from "react";

import DetailViewLink from "components/DetailViewLink"; 

/**
 * This renders a content item's details when being listed. 
 * 
 * Uses a switch to decide on particular details before actually rendering.
 * 
 * @param {object} props - The component properties from within the react context.
 */
export default function ContentItemList(props) {

    let content = props.contentData;
    let thumbnailClass;

    //its possible to limit the amount of characters on a title or description this way
    let title = content.title.substring(0, 35) + "...";
    let description = content.description.substring(0, 55) + "...";

    switch (props.pageId) {
        case "video-list":
            thumbnailClass = "thumbnail-large";
            break;
        case "side-video-list":
            thumbnailClass = "thumbnail-medium";
            break;
    }

    return (
        <div id={"content_" + content.index}>
            <div className={thumbnailClass}>
                <DetailViewLink contentId={content._id}>
                    <img src={content.thumbnailUrl} />
                </DetailViewLink>
            </div>
            <div className="title">
                <DetailViewLink contentId={content._id}>
                    {title}
                </DetailViewLink>
            </div>
            <div className="description">
                {description}
            </div>
        </div>
    );
} 