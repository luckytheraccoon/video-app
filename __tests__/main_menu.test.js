import React from "react";
import ReactDOM from "react-dom";
import {MainMenu} from '../raw_resources/js/src/components/main';
import renderer from 'react-test-renderer';

test('Main menu renders correctly', () => {
  const component = renderer.create(
    <MainMenu />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});