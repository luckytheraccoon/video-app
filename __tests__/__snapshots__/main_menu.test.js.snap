// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main menu renders correctly 1`] = `
<div
  className="div-main-menu undefined"
>
  <div
    className="div-user-block"
  >
    <div
      className="div-user-photo"
    >
      <span
        className="glyphicon glyphicon-user"
        onClick={[Function]}
      />
    </div>
    <div
      className="div-user-name"
    >
      User Name
    </div>
  </div>
  <div
    className="div-main-menu-menu"
  >
    <button>
      Menu 1
    </button>
    <button>
      Menu 2
    </button>
    <button>
      Menu 3
    </button>
  </div>
  <div
    className="div-main-menu-logout"
  >
    <button>
      Logout
    </button>
  </div>
</div>
`;
